### PR TITLE
Add initial contribution guide

### PR DESCRIPTION
**Pull request by toVersus**
Write up initial contribution guide as a basis for discussion. Feel free to add your comments!

### CODE REVIEW

**Comment by prog893** [comment]
Should this not be in template repository instead?

**Comment by toVersus** [comment]
I think we should put the link to the contributing guide in template repository, then we can easily modify the content and distribute it to the entire modules.

**Comment by prog893** [comment]
Makes sense. Will review.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Contributing
+
+This doc explains how to get started contributing to [baikonur-oss](https://github.com/baikonur-oss) project.
```
Comment (prog893) [inline CONTRIBUTING.md]: IMHO official project name should be `Baikonur`. Thus, the link should be either `Baikonur` or `Baikonur OSS` (first preferred).

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+## Filing an Issue
+
+If you are trying to use baikonur-oss modules and run into some issues, please file an issue. We can look into the issue more easily if you describe in details as possible as you can. The following is example detailed descriptions:
```
Comment (prog893) [inline CONTRIBUTING.md]: Baikonur modules, for the same reason as above.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+- Version of tools and modules
```
Comment (prog893) [inline CONTRIBUTING.md:9]: PR templates to come?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+- Expected and actual results
+- Error messages with debug option
+- Steps to reporoduce the issues
+
+## Submitting a PR
+
+Before submitting a PR, make sure that there's an issue filed for your work. It might need some discussion before addressing the issue. Filing a issue first will help us to save your time to solving misunderstandings between you and us.
```
Comment (prog893) [inline CONTRIBUTING.md]: I doubt we need this section. 

Instead, I think we should say something like this:
```
PRs are welcome! Submitting an issue and discussing is a good practice, so feel free to open new issues. Baikonur project is young, and we still have to figure out guidelines etc., so submitting an issue prior to submitting a PR might be better.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+## Releasing
+
+The release process for existing module goes something like this:
+
+1. Make sure the direction for addressing the issues under discussion
+1. Accept a PR and review immpementation
+1. Additional E2E test or something using codes within the branch
+1. Get approval and merge it into master
+1. Increce version number on **master branch**
```
Comment (prog893) [inline CONTRIBUTING.md]:
```suggestion
1. Increase version number on **master branch**
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+- Error messages with debug option
+- Steps to reporoduce the issues
+
+## Submitting a PR
+
+Before submitting a PR, make sure that there's an issue filed for your work. It might need some discussion before addressing the issue. Filing a issue first will help us to save your time to solving misunderstandings between you and us.
+
+## Releasing
+
+The release process for existing module goes something like this:
```
Comment (prog893) [inline CONTRIBUTING.md]:
```suggestion
The release process is as follows:
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+1. Make sure the direction for addressing the issues under discussion
```
Comment (prog893) [inline CONTRIBUTING.md]: I could not parse this. Please explain.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+1. Accept a PR and review immpementation
```
Comment (prog893) [inline CONTRIBUTING.md:23]:
```suggestion
1. Accept a PR and review implementation
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+1. Additional E2E test or something using codes within the branch
```
Comment (prog893) [inline CONTRIBUTING.md:24]:
```suggestion
1. Additional E2E testing  (or unit tests etc. when available) with code on the PR branch. For Lambda modules, build packages from scratch and test.
```

File: CONTRIBUTING.md (hunk 5 above)
Comment (prog893) [reply CONTRIBUTING.md]: Maybe even, `create a new incremented release from releases page`. 

Lambda packages also require a valid zip package attached with filename `lambda_package.zip`.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+1. Get approval and merge it into master
+1. Increce version number on **master branch**
+
+The release process for new module goes something like this:
+
+1. Review first implementation including sample code pointing to the Terraform Registry URI
+1. Put v0.0.1 tag on **PR branch**
+1. Register module with Terraform Registry
+1. Check the appearance of modules in Terraform Registry
+1. Additional E2E test or something using codes tagged with v0.1.0 if needed
```
Comment (prog893) [inline CONTRIBUTING.md]:
```suggestion
1. Additional E2E, unit tests (with additional v0.0.x releases if needed)
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+1. Get approval and merge it into master
+1. Put v1.0.0 tag on **master branch**
```
Comment (prog893) [inline CONTRIBUTING.md:36]: Not only tag, but we also want to have a [release](https://github.com/baikonur-oss/docs/releases).

File: DEVELOPMENT.md [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+- [Terraform](https://www.terraform.io/): Infrastructure as Code tools for automating deployment
+  - Currently, all modules are built on top of the Terraform v0.11.x, so you must pick up the latest version from [v0.11 releases](https://releases.hashicorp.com/terraform/)
+- [terraform-docs](https://github.com/segmentio/terraform-docs): A utility to generate documentation from Terraform modules
+- [pre-commit](https://pre-commit.com/): A framework for running [Git hook](https://git-scm.com/docs/githooks) automatically on every commit
+
+## Checkout the reporitory
+
+To checkout this repository:
+
+1. Create your own [folk of this repository](https://help.github.com/articles/fork-a-repo/)
```
Comment (prog893) [inline DEVELOPMENT.md]:
```suggestion
1. Create your own [fork of a repository](https://help.github.com/articles/fork-a-repo/)
```

File: DEVELOPMENT.md [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+1. Clone it into your local machine:
+
+```bash
+mkdir baikonur-oss
+cd baikonur-oss
```
Comment (prog893) [inline DEVELOPMENT.md]: These two lines seem to be unnecessary.

File: DEVELOPMENT.md [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+git clone git@github.com:${YOUR_GITHUB_USERNAME}/${BAIKONUR_MODULE_NAME}.git
+cd ${BAIKONUR_MODULE_NAME}
+git remote add upstream git@github.com:baikonur-oss/${BAIKONUR_MODULE_NAME}.git
+git remote set-url --push upstream no_push
+```
+
+We recommend to add the `upstream` remote for regulary [syncing your fork](https://help.github.com/articles/syncing-a-fork/).
+
+## Activate the pre-commit hook
```
Comment (prog893) [inline DEVELOPMENT.md]:
```suggestion
## Install pre-commit hooks
```

File: DEVELOPMENT.md [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+Activate the pre-commit hook and have the following benefits before committing the changes:
+
+- Apply terraform format
+- Automatically generate inputs/outputs doc sections based on the declarations in variables/outputs files
```
Comment (prog893) [inline DEVELOPMENT.md]:
```suggestion
- Automatically generate inputs/outputs doc sections in README based on variables, outputs declarations
```

Comment (prog893) [reply DEVELOPMENT.md]: The catch is that, for Terraform it does not matter what .tf files are named.

File: DEVELOPMENT.md [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+```bash
+pre-commit install
```
Comment (prog893) [inline DEVELOPMENT.md:41]: Must be done inside cloned repo

File: CONTRIBUTING.md (hunk 3 above)
Comment (toVersus) [reply CONTRIBUTING.md:9]: Yep. We should include a PR template in baikonur template repository. This is a temporary work.

File: CONTRIBUTING.md (hunk 7 above)
Comment (toVersus) [reply CONTRIBUTING.md]: This is related to the https://github.com/baikonur-oss/docs/pull/1#discussion_r305209300. I think we should discuss how to resolve the issue in filed Issue before accepting a PR.

Comment (prog893) [reply CONTRIBUTING.md]: I personally do not think we **must** have issues for PRs at all times.

Can we remove/comment out this line temporarily until we sort this out?

**Review by prog893 — APPROVED** [review]
Looks good for a first version

**Pull request merged**